Adds 3DTILES_bounding_volume_S2 extension (#439)
CesiumGS/3d-tiles: extensions/3DTILES_bounding_volume_S2/0.0.0/README.md

# 3DTILES_bounding_volume_S2

## Contributors

> 

Reviewer: Maybe @ptrgags, you, and I?

## Status

Draft

## Dependencies

Written against 3D Tiles 1.0. It may be used in conjunction with [`3DTILES_implicit_tiling`](https://github.com/CesiumGS/3d-tiles/tree/3d-tiles-next/extensions/3DTILES_implicit_tiling/0.0.0).

## Overview

[S2](http://s2geometry.io/) is a spherical geometry library that represents all data on a 3D sphere, unlike traditional libraries that use 2D planar projections. This makes it possible to represent the globe with no seams or singularities, with low distortion everywhere on Earth. The S2 library involves projecting the 6 faces of a cube on a unit sphere, creating 6 root "cells" that subdivide evenly into 4 tiles, into a quadtree structure.

This extension to 3D Tiles enables using S2 cells as a `boundingVolume`, and uses `3DTILES_implict_tiling` to enforce subdivision.

| S2 Curve on Cube Face  |  S2 Curve on WGS84 Ellipsoid |
|---|---|
| ![Math](figures/plane.png)  | ![Math](figures/ellipsoid.png)  |

## Coordinate System

The S2 library does not mandate the usage of geocentric or geodetic coordinates, however, for this extension, we use WGS84 geodetic coordinates for mapping the points between the Earth and the S2 sphere.

## Cell Token

> Cells in S2 are repesented using 64-bit `S2CellId`s. To ensure these numbers are not interpreted as 32-bit integers, in this extension, they are required to be encoded in their hexadecimal string representation, the `S2CellToken`. More details on computing an `S2CellToken` can be found in the [S2 reference implementation](https://github.com/google/s2-geometry-library-java/blob/c28f287b996c0cedc5516a0426fbd49f6c9611ec/src/com/google/common/geometry/S2CellId.java#L468).

Reviewer: > To ensure these numbers are not interpreted as 32-bit integers

This might not be clear enough. It's less about being interpreted as 32 bit and more about languages not supporting 64-bit. Ignoring `BigInt` for now, JS can safely represent integers up to `2^53 - 1` ([`Number.MAX_SAFE_INTEGER`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number/MAX_SAFE_INTEGER)), unless there's bit shifting involved in which case it is limited to 32 bit integers.

## Heights

The S2 cell itself is used to specify an area on the surface of the ellipsoid. To create a bounding volume, the `minumumHeight` and `maximumHeight` properties must be specified. These heights must be specified in meters above the WGS84 ellipsoid.

| Parent Cell  |  Children Cells |
|---|---|
| ![Math](figures/parent.png)  | ![Math](figures/children.png)  |

> ## Subdivision

Reviewer: The spec text should include an example of what S2 + implicit looks like.

Author: Good thing you suggested this because it made me add a note about `maximumLevel` not being greater than `30 - {Level of root S2 cell}`.

The S2 library defines a [cell hierarchy](http://s2geometry.io/devguide/s2cell_hierarchy), that follows uniform subdivision using a quadtree structure, where each cell subdivides into 4 smaller cells that combine to occupy the same area as the parent.

> When used with `3DTILES_implicit_tiling`, a `quadtree` subdivision scheme will follow the rules for subdivision as defined by S2. When an `octree` subdivision scheme is used, the split in the vertical dimension occurs at the midpoint of the `minimumHeight` and `maximumHeight` of the parent tile.

Reviewer: ```suggestion
When used with `3DTILES_implicit_tiling`, a `QUADTREE` subdivision scheme will follow the rules for subdivision as defined by S2. When an `OCTREE` subdivision scheme is used, the split in the vertical dimension occurs at the midpoint of the `minimumHeight` and `maximumHeight` of the parent tile.
```

> When used with `3DTILES_implicit_tiling`, a `quadtree` subdivision scheme will follow the rules for subdivision as defined by S2. When an `octree` subdivision scheme is used, the split in the vertical dimension occurs at the midpoint of the `minimumHeight` and `maximumHeight` of the parent tile.

Reviewer: It's not completely clear from this section how tile availability, which uses a morton curve, maps to S2 cell IDs, which use a hilbert curve.

> When used with `3DTILES_implicit_tiling`, a `QUADTREE` subdivision scheme will follow the rules for subdivision as defined by S2. When an `OCTREE` subdivision scheme is used, the split in the vertical dimension occurs at the midpoint of the `minimumHeight` and `maximumHeight` of the parent tile.

Reviewer 2: A diagram would be helpful here on how the octree looks.

## Schema Changes

> The changes to the schema can documented in [extension.schema.json](schema/extension.schema.json).

Reviewer: The schema file should be renamed to `boundingVolume.3DTILES_bounding_volume_S2.schema.json`

> The changes to the schema can documented in [extension.schema.json](schema/extension.schema.json).

Reviewer 2: (commenting about one of the images) -- I see that `math.png` is still in the file list, is that needed anymore?

- Sean Lilley, Cesium
- Peter Gagliardi, Cesium

<!-- omit in toc -->
## Status

Draft

<!-- omit in toc -->
## Dependencies

Written against 3D Tiles 1.0.

<!-- omit in toc -->
## Optional vs. Required

This extension is required, meaning it must be placed in both the `extensionsUsed` and `extensionsRequired` lists in the tileset JSON.

<!-- omit in toc -->
## Contents

- [3DTILES_bounding_volume_S2](#3dtiles_bounding_volume_s2)
  - [Overview](#overview)
  - [Coordinate System](#coordinate-system)
  - [Bounding Volume](#bounding-volume)
    - [Cell Token](#cell-token)
    - [Bounding Heights](#bounding-heights)
  - [Subdivision](#subdivision)
  - [Property Reference](#property-reference)
  - [Schema Changes](#schema-changes)

## Overview

[S2](http://s2geometry.io/) is a spherical geometry library that maps the 6 faces of a unit cube onto the unit sphere. Typically, traditional GIS libraries use planar projections to map data between a 2D plane and a 3D ellipsoid (representing the Earth). Since a sphere is a closer approximation of the shape of the ellipsoid, S2 makes it possible to represent the globe with no seams or singularities, with low distortion everywhere on Earth.

> |Level 0|Level 1|

Reviewer 2: Looking at these diagrams in GH, I like the overall look, but wish that A. the view was a little bit more zoomed in so the globe fills more of the frame and B. that the text is bigger and more readable.

![image](https://user-images.githubusercontent.com/8422414/118716599-02fcf400-b7f3-11eb-8b68-ab129ca36435.png)

|:-:|:-:|

> | ![Plane - Level 0](figures/plane_root.png) S2 cell (`"1"`) on the Earth cube| ![Plane - Level 1](figures/plane_children.png) Children of S2 cell (`"1"`) on the Earth cube |

Reviewer 2: Just a newbie S2 question: why are the 4 children `04, 0c, 14, 1c`? I would have thought they'd all start with 1 if this is face 1, but maybe S2 works differently than I think?

| ![Ellipsoid - Level 0](figures/ellipsoid_root.png)  S2 cell (`"1"`) on the WGS84 ellipsoid| ![Ellipsoid - Level 1](figures/ellipsoid_children.png) Children of S2 cell (`"1"`) on the WGS84 ellipsoid|

In S2, each face of the unit cube can be subdivided into 30 levels using a quadtree structure, in which each "cell" or tile on the grid subdivides into 4 equal cells or tiles at the subsequent level. This extension to 3D Tiles enables using these cells as the basis for bounding volumes for tiles.

| S2 Curve on Earth cube |  S2 Curve on WGS84 ellipsoid |
|:-:|:-:|
| ![Math](figures/plane.png)  | ![Math](figures/ellipsoid.png)  |

> The S2 library uses a modified Hilbert curve to provide a one dimensional ordering of cells on the S2 Earth cube. This provides each cell, from level 1 to level 30, with a unique 64-bit identifier. Therefore, we can uniquely identify millimeter scale areas on Earth with their S2 cell IDs.

Reviewer 2: I would use passive voice here instead of "we". Also the "Therefore" doesn't make sense in this sentence, it's not immediately obvious that 64-bit IDs imply mm precision

```suggestion
The S2 library uses a modified Hilbert curve to provide a one dimensional ordering of cells on the S2 Earth cube. This provides each cell, from level 1 to level 30, with a unique 64-bit identifier. Areas as small as a millimeter can be identified uniquely using S2 cell IDs.
```

## Coordinate System

The S2 library does not mandate the usage of geocentric or geodetic coordinates. This extension uses WGS84 geodetic coordinates for mapping the points between the Earth and the S2 sphere.

## Bounding Volume

This extension enables using S2 cells and a set of bounding heights to describe a bounding volume in 3D Tiles. The `token` represents an S2 cell. The cell describes 4 vertices on the surface of the WGS84 ellipsoid for the area, and the `minimumHeight` and `maximumHeight` describe the vertical limits with respect to the WGS84 ellipsoid.

> ![Volume](figure/../figures/volume.jpg)

Reviewer 2: Looks like you checked in a `volume.png` that looks unused, could you remove this?

> ![Volume](figure/../figures/volume.jpg)

Reviewer 2: I like the layout of this diagram, but could the bounding volume have a color that contrasts the city? I think that would make it easier to understand.

Reviewer 2: Oh wow the vertices are numbered? that was hard to read. Also not sure why we need to number them?

<center>S2 cell (<tt>"89c6c7"</tt>) covering the Philadelphia Center City area, with minimum height set to <tt>0</tt> m and maximum height set to <tt>1000</tt> m.</center>

### Cell Token

This extension uses tokens, or hexadecimal string representations of S2 cell identifier for two reasons:
 1. Precision: Using a token will require a client to convert it to the correct data type: `uint64`

>  2. Conciseness: Tokens provide a more concise representation of identifiers

Reviewer 2: Could you explain what you mean by "more concise representation"? Not sure I'm following.

Author: I added an example to show, not just tell:
>  For example, the root (level 0) cell with cell ID `3458764513820540928` has the token `"3"`.

The S2 cell itself is used to specify an area on the surface of the ellipsoid. To create a bounding volume, the `minimumHeight` and `maximumHeight` properties must be specified. These heights must be specified in meters above the WGS84 ellipsoid.

## Subdivision

The S2 library defines a [cell hierarchy](http://s2geometry.io/devguide/s2cell_hierarchy), that follows uniform subdivision using a quadtree structure, where each cell subdivides into 4 smaller cells that combine to occupy the same area as the parent.

When used with `3DTILES_implicit_tiling`, a `QUADTREE` subdivision scheme will follow the rules for subdivision as defined by S2. When an `OCTREE` subdivision scheme is used, the split in the vertical dimension occurs at the midpoint of the `minimumHeight` and `maximumHeight` of the parent tile. The `availability` bitstreams are ordered by the Morton index of the tile, as specified by `3DTILES_implicit_tiling`, not by the Hilbert index used by S2. Additionally, the `maximumLevel` property cannot be greater than `30 - {Level of root S2CellId}` because S2 cell hierarchy only extends to level 30.

| Cell  | Quadtree Subdivision | Octree Subdivision |
|---|---|---|

> | ![Parent Cell](figures/parent.png)  | ![Quadtree Cells](figures/quadtree.png)  | ![Octree Cells](figures/octree.png)  |

Reviewer 2: For octree.png, I would try to find a viewpoint that emphasizes that this is a 3D division, at first glance this looked like a quadtree to me.

}
```

## Property Reference

**`3DTILES_bounding_volume_S2 extension` Properties**

|   |Type|Description|Required|
|---|---|---|---|
|**token**|`string`|A hexadecimal representation of the S2 cell identifier. Tokens must be lower-case (except X), must not contain whitespace and must have trailing zeros stripped.| &#10003; Yes|
|**minimumHeight**|`number`|The minimum height of the tile, specified in meters above (or below) the WGS84 ellipsoid.| &#10003; Yes|
|**maximumHeight**|`number`|The maximum height of the tile, specified in meters above (or below) the WGS84 ellipsoid.| &#10003; Yes|
|**extensions**|`object`|Dictionary object with extension-specific objects.| No|
|**extras**|`any`|Application-specific data.| No|

Additional properties are not allowed.

<!-- omit in toc -->
#### token ✅

> A hexadecimal representation of the S2CellId. The 0 S2CellId must be represented as 'X'. Tokens must be lower-case (except X), must not contain whitespace and must have trailing zeros stripped.

Reviewer 2: Are we allowing X as a valid token? If so, explain what it means. Otherwise, maybe disallow it?

* **Type**: `string`
* **Required**:  &#10003; Yes

<!-- omit in toc -->
#### minimumHeight ✅

The minimum height of the tile, specified in meters above (or below) the WGS84 ellipsoid.

* **Type**: `number`
* **Required**:  &#10003; Yes

<!-- omit in toc -->
#### maximumHeight ✅

The maximum height of the tile, specified in meters above (or below) the WGS84 ellipsoid.

* **Type**: `number`
* **Required**:  &#10003; Yes

<!-- omit in toc -->
#### extensions

Dictionary object with extension-specific objects.

* **Type**: `object`
* **Required**: No

<!-- omit in toc -->
#### extras

Application-specific data.

* **Type**: `any`
* **Required**: No

> ## Schema Changes

Reviewer 2: Did we use "Schema Changes" in other specs? It seems a bit odd to me, you're really defining a new schema, not modifying an existing one.
